[sdk/python] - AssertionError: Unexpected type. Expected 'dict' got '<class 'str'>' for Policy

<!-- Please provide a general summary of the issue. -->
Before 3.0.0 I could pass in a dict as an `aws.BucketPolicy.policy`. Now the same code fails with: `AssertionError: Unexpected type. Expected 'dict' got '<class 'str'>'`

## Expected behavior
<!-- Please tell us what should happen -->

The code should succeed as it did before

## Current behavior
<!-- Please tell us what happens instead of the expected behavior -->

It fails with an assertion error

## Steps to reproduce
<!-- Provide a link to a live example, or an unambiguous set of steps to reproduce this bug. Include code to reproduce, if relevant. -->

The following code reproduces the issue if you try to run `pulumi up`

```python
import pulumi
from pulumi_aws import s3

bucket = s3.Bucket("my-bucket")

policy = s3.BucketPolicy("my-policy", bucket=bucket.id, policy={
    "Version": "2012-10-17",
    "Statement": {
        "Effect": "Allow",
        "Principal": "*",
        "Action": ["s3:GetObject"],
        # Policy refers to bucket explicitly
        "Resource": [pulumi.Output.concat("arn:aws:s3:::", bucket.id, "/*")]
    }
})
```

## Context (Environment)
<!-- How has this issue affected you? What are you trying to accomplish? Providing context helps us come up with a solution that is most useful in the real world -->

## Affected feature

<!-- If you know that your bug is happening in a specific feature, please list it here. And, if you see a matching label with the `area/` prefix, please feel free to apply it. -->

I can repro the S3 example. With some debug stmts:

```
    error: Traceback (most recent call last):
      File "/Users/anton/.nix-profile/bin/pulumi-language-python-exec", line 85, in <module>
        loop.run_until_complete(coro)
      File "/nix/store/cp9nlaz4a24n8sp9f738ngdnrd19s4rp-python3-3.8.5/lib/python3.8/asyncio/base_events.py", line 616, in run_until_complete
        return future.result()
      File "/Users/anton/pulumi/sdk/python/lib/pulumi/runtime/stack.py", line 110, in run_in_stack
        await run_pulumi_func(lambda: Stack(func))
      File "/Users/anton/pulumi/sdk/python/lib/pulumi/runtime/stack.py", line 45, in run_pulumi_func
        await wait_for_rpcs()
      File "/Users/anton/pulumi/sdk/python/lib/pulumi/runtime/stack.py", line 94, in wait_for_rpcs
        raise exception
      File "/Users/anton/pulumi/sdk/python/lib/pulumi/runtime/rpc_manager.py", line 67, in rpc_wrapper
        result = await rpc
      File "/Users/anton/pulumi/sdk/python/lib/pulumi/output.py", line 91, in is_value_known
        return await is_known and not contains_unknowns(await future)
      File "/Users/anton/pulumi/sdk/python/lib/pulumi/output.py", line 91, in is_value_known
        return await is_known and not contains_unknowns(await future)
      File "/Users/anton/pulumi/sdk/python/lib/pulumi/output.py", line 91, in is_value_known
        return await is_known and not contains_unknowns(await future)
      [Previous line repeated 1 more time]
      File "/Users/anton/pulumi/sdk/python/lib/pulumi/runtime/resource.py", line 438, in do_register
        resolver = await prepare_resource(res, ty, custom, remote, props, opts, typ)
      File "/Users/anton/pulumi/sdk/python/lib/pulumi/runtime/resource.py", line 100, in prepare_resource
        serialized_props = await rpc.serialize_properties(props, property_dependencies_resources, translate, typ)
      File "/Users/anton/pulumi/sdk/python/lib/pulumi/runtime/rpc.py", line 121, in serialize_properties
        result = await serialize_property(v, deps, input_transformer, get_type(k))
      File "/Users/anton/pulumi/sdk/python/lib/pulumi/runtime/rpc.py", line 307, in serialize_property
        raise AssertionError(f"Unexpected type. Expected 'dict' got '{typ}'\n" + m)
    AssertionError: Unexpected type. Expected 'dict' got '<class 'str'>'
    serialize_property(value={'Version': '2012-10-17', 'Statement': {'Effect': 'Allow', 'Principal': '*', 'Action': ['s3:GetObject'], 'Resource': [<pulumi.output.Output object at 0x10f188ca0>]}}, deps=[], input_transformer=None, typ=<class 'str'>
     type(value) = <class 'dict'>
```

--- a/CHANGELOG_PENDING.md
+++ b/CHANGELOG_PENDING.md
@@ -45,6 +45,9 @@
 - [sdk/python] Fix type-related regression on Python 3.6.
   [#6942](https://github.com/pulumi/pulumi/pull/6942)
 
+- [sdk/python] Don't error when a dict input value has a mismatched type annotation.
+  [#6949](https://github.com/pulumi/pulumi/pull/6949)
+
 ### Misc.
 
 - [auto/dotnet] Bump YamlDotNet to 11.1.1

--- a/sdk/python/lib/pulumi/runtime/rpc.py
+++ b/sdk/python/lib/pulumi/runtime/rpc.py
@@ -300,7 +300,15 @@ async def serialize_property(value: 'Input[Any]',
                         get_type = lambda k: args[1]
                         translate = None
                 else:
-                    raise AssertionError(f"Unexpected type. Expected 'dict' got '{typ}'")
+                    translate = None
+                    # Note: Alternatively, we could assert here that we expected a dict type but got some other type,
+                    # but there are cases where we've historically allowed a user-defined dict value to be passed even
+                    # though the type annotation isn't a dict type (e.g. the `aws.s3.BucketPolicy.policy` input property
+                    # is currently typed as `pulumi.Input[str]`, but we've allowed a dict to be passed, which will
+                    # "magically" work at runtime because the provider will convert the dict value to a JSON string).
+                    # Ideally, we'd update the type annotations for such cases to reflect that a dict could be passed,
+                    # but we haven't done that yet and want these existing cases to continue to work as they have
+                    # before.
 
         obj = {}
         # Don't use value.items() here, as it will error in the case of outputs with an `items` property.

--- a/sdk/python/lib/test/langhost/input_type_mismatch/__init__.py
+++ b/sdk/python/lib/test/langhost/input_type_mismatch/__init__.py
@@ -1,0 +1,13 @@
+# Copyright 2016-2021, Pulumi Corporation.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.

--- a/sdk/python/lib/test/langhost/input_type_mismatch/__main__.py
+++ b/sdk/python/lib/test/langhost/input_type_mismatch/__main__.py
@@ -1,0 +1,46 @@
+# Copyright 2016-2021, Pulumi Corporation.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import pulumi
+
+@pulumi.input_type
+class MyResourceArgs:
+    def __init__(self, policy: pulumi.Input[str]):
+        pulumi.set(self, "policy", policy)
+
+    @property
+    @pulumi.getter
+    def policy(self) -> pulumi.Input[str]:
+        return pulumi.get(self, "policy")
+
+class MyResource(pulumi.CustomResource):
+    def __init__(self, name, args: MyResourceArgs):
+        super().__init__("test:index:MyResource", name, args)
+
+    @property
+    @pulumi.getter
+    def policy(self) -> pulumi.Output[str]:
+        return pulumi.get(self, "policy")
+
+r1 = MyResource("testResource", MyResourceArgs(policy='{"hello": "world"}'))
+
+# We have cases where an input is typed as `Input[str]` but passing a dict still works despite
+# the type annotation not saying that it accepts a dict (e.g. aws.s3.BucketPolicy.policy, which
+# is typed as `Input[str]` but passing a dict works because the provider will convert the dict
+# to a JSON string). We need to fix the type annotations in these cases, but in the meantime such
+# cases should continue to work as before without an error being raised by the SDK.
+r2 = MyResource("testResource", MyResourceArgs(policy={"hello": "world"}))
+
+pulumi.export("r1.policy", r1.policy)
+pulumi.export("r2.policy", r2.policy)

--- a/sdk/python/lib/test/langhost/input_type_mismatch/test_input_type_mismatch.py
+++ b/sdk/python/lib/test/langhost/input_type_mismatch/test_input_type_mismatch.py
@@ -1,0 +1,46 @@
+# Copyright 2016-2021, Pulumi Corporation.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import json
+from os import path
+from ..util import LanghostTest
+
+
+class InputTypeMismatchTest(LanghostTest):
+    def test_input_type_mismatch(self):
+        self.run_test(
+            program=path.join(self.base_path(), "input_type_mismatch"),
+            expected_resource_count=2)
+
+    def register_resource(self, ctx, dry_run, ty, name, _resource,
+                          _dependencies, _parent, _custom, _protect, _provider, _property_deps, _delete_before_replace,
+                          _ignore_changes, _version):
+        self.assertEqual("test:index:MyResource", ty)
+
+        policy = _resource["policy"]
+        if isinstance(policy, dict):
+            policy = json.dumps(policy)
+
+        return {
+            "urn": self.make_urn(ty, name),
+            "id": name,
+            "object": {"policy": policy},
+        }
+
+    def register_resource_outputs(self, _ctx, _dry_run, _urn, ty, _name, _resource, outputs):
+        self.assertEqual("pulumi:pulumi:Stack", ty)
+        self.assertEqual({
+            "r1.policy": '{"hello": "world"}',
+            "r2.policy": '{"hello": "world"}',
+        }, outputs)

